---
title: "QBio305 - Genetic analysis of phenotypic variation measured in the Oregon Wolfe Barley collection"
author: "Ingo Giebel"
format: html
   toc: true
server: shiny
---

```{r}
#| label: load-packages
#| echo: false

# For determining the project root
library(here)
# For transforming and representing data
library(tidyverse)
# For plotting
library(ggpubr)
# For Levene's test
library(car)
# For linkage analysis
library(qtl)
```

```{r}
#| label: file-paths
#| echo: false

dta_owb_genetic_map_file_path <- here::here(
  "data",
  "dta_owb_genetic_map.csv"
)

dta_owb_height_and_spike_length_by_genotype_file_path <- here::here(
  "data",
  "dta_owb_height_and_spike_length_by_genotype.csv"
)

dta_owb_linkage_map_qualt_phenotypes_file_path <- here::here(
  "data",
  "dta_owb_linkage_map_qualt_phenotypes.csv"
)
```

```{r}
#| label: r-options
#| echo: false

options(scipen = 1)
```

```{r}
#| label: func-defs
#| echo: false

# Plot a histogram of the residuals
plot_residuals_hist <- function(data_aov, xlab) {
  res <- data_aov$residuals
  ggplot2::ggplot(
    data = data.frame(
      1:length(res),
      res
    ),
    mapping = ggplot2::aes(x = res),
  ) +
    ggplot2::geom_histogram(
      mapping = ggplot2::aes(y = ggplot2::after_stat(width * density)),
      color = "darkgrey",
      bins = 50
    ) +
    ggplot2::stat_function(
      fun = dnorm,
      args = list(
        mean = mean(res),
        sd = sd(res)
      ),
      color = "darkred"
    ) +
    ggplot2::labs(
      x = xlab,
      y = "Density"
    )
}

# Conduct an ANOVA of phenotype ~ marker for each marker within marker_names
anova_all_markers <- function(data, genetic_map, phenotype, marker_names) {
  # Tibble for the markers, theirs positions and significance values for the
  # phenotype
  marker_pval_df <- tibble::tibble(
    marker = character(),
    chr = character(),
    pos = numeric(),
    pval = numeric()
  )

  # Loop through all markers and conduct ANOVAs for each one of them
  for (marker_name in marker_names) {
    # Extract the marker's genetic position from the genetic map
    marker_chr_pos <- dplyr::filter(genetic_map, marker == marker_name)
    # Conduct an one-way ANOVA of the plant height vs the marker
    marker_aov <- aov(
      as.formula(paste(phenotype, "~", marker_name)),
      data = data
    )
    # Extract the marker's p-value from the ANOVA table
    pval <- summary(marker_aov)[[1]][1, 5]
    # Add a new row for the marker, its position, and its p-value
    marker_pval_df <- marker_pval_df |> tibble::add_row(marker_chr_pos, pval)
  }

  marker_pval_df
}

# Determine the LOD scores of the given marker for all possible positions
calc_load_scores <- function(linkage_map, phenotype) {
  # Try all possible positions for the marker, keeping all other markers
  # fixed, and evaluate the log likelihood and estimate the chromosome length
  qtl::tryallpositions(
    linkage_map,
    marker = phenotype,
    verbose = FALSE
  )
}

# Map the qualitative marker to the most likely locus responsible for
# the trait. Returns an updated linkage map.
map_qualt_trait <- function(linkage_map, phenotype, lod_scores) {
  # Determine the LOD entry with the highes LOD score
  best_marker_pos <- dplyr::slice_max(lod_scores, n = 1, lod)

  # Move the phenotype marker to the position with the highest LOD score
  qtl::movemarker(
    linkage_map,
    marker = phenotype,
    newchr = dplyr::select(best_marker_pos, chr)[1, 1],
    newpos = dplyr::select(best_marker_pos, pos)[1, 1]
  )
}
```

```{r}
#| label: read-data
#| echo: false

# Read in the raw quantitative data of the different OWB genotypes
dta_quant_raw <- car::strings2factors(
  readr::read_csv(dta_owb_height_and_spike_length_by_genotype_file_path),
  verbose = FALSE
)

# Read in the genetic map of OWB
dta_owb_genetic_map <- readr::read_csv(
  dta_owb_genetic_map_file_path
)

# Read in linkage map and qualitative phenotypes as denoted in marker annotation
dta_linkage_map_qualt_phenotypes <- qtl::read.cross(
  format = "csv",
  file = dta_owb_linkage_map_qualt_phenotypes_file_path,
  genotypes = c("a", "b"),
  alleles = c("a", "b"),
  crosstype = "dh"
)
```

```{r}
#| label: anova-raw-height-genotype
#| echo: true

# Densitiy plot comparing the height of the different genotypes
ggpubr::ggdensity(
  data = dta_quant_raw,
  x = "Height",
  add = "mean",
  color = "Genotype",
  fill = "Genotype",
  xlab = "Height in cm",
  ylab = "Density",
) +
  ggplot2::theme(legend.justification = c("left", "top")) +
  ggplot2::guides(
    colour = ggplot2::guide_legend(
      title.position = "top",
      nrow = 6,
      byrow = TRUE
    )
  )

# Summary statistics per genotype
dta_quant_raw |>
  dplyr::select(Genotype, Height) |>
  dplyr::group_by(Genotype) |>
  dplyr::summarise(
    Count = n(),
    Mean = mean(Height),
    Sd = sd(Height)
  )

# Fit an analysis of variance model for the height vs genotype
height_raw_aov <- aov(Height ~ Genotype, data = dta_quant_raw)

# Summary of the ANOVA analysis
summary(height_raw_aov)

# Conclusion: The genotype has a significant effect on the plant height
# (p-value < 2e-16).
```

```{r}
#| label: anova-raw-check-homogeneity-variance
#| echo: true

# ANOVA assumes that the data are normally distributed and that the
# variance across the groups is approximately equal. These assumptions
# are assessed with some diagnostic plots and tests.

# Plot the residuals vs fit in order to check the homogeneity of variance
# assumption
plot(height_raw_aov, which = 1)

# Levene's test for homogeneity of variance across groups
car::leveneTest(Height ~ Genotype, data = dta_quant_raw)

# Conclusion: the variances are roughly homogeneous. The p-value for the
# Levene's test of the raw data is 0.7938. This means that the variances
# within the groups are not significantly different from each other, i.e.,
# the homogeneity assumption of the variance is met.

# Outliers: the rows 14, 40, and 182 are detected as outliers, which can
# severely affect normality and homogeneity of the variance. Therefore, these
# outliers (in the genotypes OWB33, OWB76, and OWB54) will be removed to meet
# the test assumptions.
```

```{r}
#| label: anova-raw-check-normality
#| echo: true

# Plot a histogram of the residuals
plot_residuals_hist(height_raw_aov, "Residuals of the ANOVA height ~ genotype")

# Q-Q plot of the ANOVA residuals
plot(height_raw_aov, which = 2)

# Cook's distance plot
plot(height_raw_aov, which = 4)

# Shapiro-Wilk test for normality
shapiro.test(height_raw_aov$residuals)

# Conclusion: According to the Shapiro-Wilk test for normality, it is unlikely
# that the raw data is normally distributed (p-values < 2.2e-16). Nevertheless,
# the plots reveal that the data conform at least roughly to a normal
# distribution, in particular without the three outliers.
# Therefore, the further analysis will be done using ANOVA with cleansed data
# (without the outliers).
```

```{r}
#| label: anova-clean-height-genotype
#| echo: true

# Remove outliers
dta_quant_clean <- dta_quant_raw[-c(14, 40, 182), ]

# Fit an analysis of variance model for the height (cleansed data)
height_clean_aov <- aov(Height ~ Genotype, data = dta_quant_clean)

# Summary of the ANOVA analysis
summary(height_clean_aov)

# Plot a histogram of the residuals
plot_residuals_hist(
  height_clean_aov,
  "Residuals of the ANOVA height ~ genotype"
)

# Q-Q plot of the ANOVA residuals
plot(height_clean_aov, which = 2)

# Cook's distance plot
plot(height_clean_aov, which = 4)

# Shapiro-Wilk test for normality
shapiro.test(height_clean_aov$residuals)

# p-value = 7.823e-14
# Still very unlikely, but much more likely than for the raw data.

# Conclusion: According to the Shapiro-Wilk test for normality, it is still
# unlikely that the cleansed data is normally distributed (p-value = 7.823e-14).
# Nevertheless, the plots reveal that the data conform at least roughly to a
# normal distribution, in particular without the removed three outliers.
```

```{r}
#| label: anova-clean-height-markers
#| echo: true

marker_pval_df <- anova_all_markers(
  dta_quant_clean,
  dta_owb_genetic_map,
  "Height",
  colnames(dta_quant_clean)[5:ncol(dta_quant_clean)]
)

# Calculate the Bonferroni corrected significance threshold. The Bonferroni
# correction adjusts the chosen significance level (here 5 %) by dividing it
# by the number of tests.
sig_thld_bonfcorr <- 0.05 / nrow(marker_pval_df)

# Create a data frame that follows the requirements of the
# Manhattan command of the qqman library
marker_pval_manhattan_plot_df <- marker_pval_df |>
  dplyr::mutate(CHR = as.numeric(gsub("H", "", chr))) |>
  dplyr::rename(BP = pos) |>
  dplyr::mutate(SNP = 1:n(), .before = pval) |>
  dplyr::select(CHR, BP, SNP, pval)

# Manhattan plot of the genome-wide p-values for the markers
qqman::manhattan(
  marker_pval_manhattan_plot_df,
  genomewideline = -log(sig_thld_bonfcorr),
  suggestiveline = FALSE,
  logp = TRUE,
  p = "pval",
  lwd = 3,
  ylab = "p-value (neg. log)"
)

# Determine all gene positions which have p-value less than the Bonferroni
# corrected p-value
marker_pval_sig_df <- marker_pval_df |>
  dplyr::filter(pval < sig_thld_bonfcorr) |>
  dplyr::rename(Chromosome = chr) |>
  dplyr::group_by(Chromosome) |>
  dplyr::summarise(
    Count = n(),
    "Min. p-value" = min(pval),
    "Median p-value" = median(pval),
    "Min. chr position" = min(pos),
    "Max. chr position" = max(pos)
  )
marker_pval_sig_df

# Result:

# Chr. | Count | Min. p-val | Median p-val | Min. chr pos | Max. chr pos

# 1H      37     1.74e- 7     4.24e- 6       52.7           157.
# 2H     180     1.10e-42     1.82e-14       1.26           187.
# 3H      68     5.43e- 8     2.57e- 6       58.6           146.
# 5H      38     1.48e- 8     4.45e- 7       137.           238.
# 6H       1     6.37e- 6     6.37e- 6       129.           129.

# Conclusion: The plant height is mainly significantly effected by
# a genomic region within gene 2. But there are also markers within
# the genes 1H, 3H, 5H, and 6H which have a significant impact on
# the plant height.
```

```{r}
#| label: map-leaf-variegation-trait
#| echo: true

# Determine the LOD scores of the leaf variegation for all possible positions
load_scores <- calc_load_scores(
  dta_linkage_map_qualt_phenotypes,
  "leaf_variegation"
)
load_scores

# Map the leaf variegation marker to the most likely locus responsible
# for the trait
dta_linkage_map_qualt_phenotypes <- map_qualt_trait(
  dta_linkage_map_qualt_phenotypes,
  "leaf_variegation",
  load_scores
)

# Plot the updated linkage map
qtl::plotMap(dta_linkage_map_qualt_phenotypes,
  main = "",
  show.marker.names = TRUE
)

# Conclusion: The leaf variegation of the Oregon Wolfe barley collection
# is significantly effected by the chromosome 2H at position 192.8.

# Remark: According to
#
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8066651/
#
# the leaf variegation is controlled by the gene Wst. Therefore, our findings
# are in accordance with the findings of that paper.
# See figure
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8066651/figure/plants-10-00694-f003/
# for an illustration of the Wst locus in gene 2H.
```
