\section{Results}

\subsection{Analysis of variance for the genotype effect on the plant height}

To investigate the association between the height of the plants and their genotype, we conducted a single-factor analysis of variance (ANOVA). The analysis revealed a highly significant impact of the genotype on the plant height (p-value < \num{2e-16} for both the raw and the cleansed data in which three outliers were removed). As the corresponding F-value for the cleansed data (\num{834.4}) is substantially greater than the corresponding F-value for the raw data (\num{198.1}), the actual p-value for the cleansed data is actually even much smaller than the p-value for the raw data. Figure \ref{fig:raw_genotype_height} shows density plots of the raw height data for the different genotypes.

\begin{figure}[htbp]
    \includesvg[width=\linewidth]{../results/raw_genotype_height}
    \caption{Density plots of the raw height data for the different genotypes}
    \label{fig:raw_genotype_height}
\end{figure}

\subsection{ANOVA homogeneity of variance assumption}

ANOVA assumes that the data are normally distributed and that the variance across the groups is approximately equal. These assumptions were assessed with some diagnostic plots and tests. Figure \ref{fig:raw_aov_height_residuals_vs_fit} shows the ANOVA residuals (raw data) versus the fitted values. This plot clearly shows three outliers (in the genotypes OWB33, OWB76, and OWB54) which were later removed in the cleansed data because they severely affected the normality of the variance.

\begin{figure}[htbp]
    \includesvg[width=\linewidth]{../results/raw_aov_height_residuals_vs_fit}
    \caption{ANOVA residuals vs fitted values. Used for visual assessment of the homogeneity of variance assumption and to identify any outliers.}
    \label{fig:raw_aov_height_residuals_vs_fit}
\end{figure}

Furthermore, a Levene's test for homogeneity of variance was done. The p-value for that test was \num{0.7938}, which means that the variances across the groups were not significantly different from each other, i.e., the homogeneity assumption of the variance was met.

\subsection{ANOVA normality of variance assumption}

The normality of variance was assessed for both the raw data and the cleansed data in which the three most extreme outliers were removed. Figures \ref{fig:aov_height_residuals_histogram}, \ref{fig:aov_height_residuals_qq}, and \ref{fig:aov_height_residuals_cooks_dist} show a visual comparison of the raw and the cleansed data: a histogram of the ANOVA residuals, a Q-Q plot of the ANOVA residuals, and a Cook's distance plot of the ANOVA residuals, respectively.

Furthermore, a Shapiro-Wilk test for normality was done for both the raw and the cleansed data. According to the p-values of the Shapiro-Wilk test (raw data < \num{2.2e-16}, cleansed data \num{7.823e-14}), both the raw and the cleansed data was not normally distributed in a narrow sense. Nevertheless, the plots reveal that the data conform at least roughly to a normal distribution, in particular the cleansed data without the three most extreme outliers.

\begin{figure}[htbp]
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/raw_aov_height_residuals_histogram}
        \caption{Raw data}
        \label{fig:raw_aov_height_residuals_histogram}
    \end{subfigure}
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/clean_aov_height_residuals_histogram}
        \caption{Cleansed data}
        \label{fig:clean_aov_height_residuals_histogram}
    \end{subfigure}
    \caption{Histogram of the ANOVA residuals}
    \label{fig:aov_height_residuals_histogram}
\end{figure}

\begin{figure}[htbp]
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/raw_aov_height_residuals_qq}
        \caption{Raw data}
        \label{fig:raw_aov_height_residuals_qq}
    \end{subfigure}
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/clean_aov_height_residuals_qq}
        \caption{Cleansed data}
        \label{fig:clean_aov_height_residuals_qq}
    \end{subfigure}
    \caption{Q-Q plot of the ANOVA residuals}
    \label{fig:aov_height_residuals_qq}
\end{figure}

\begin{figure}[htbp]
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/raw_aov_height_residuals_cooks_dist}
        \caption{Raw data}
        \label{fig:raw_aov_height_residuals_cooks_dist}
    \end{subfigure}
    \begin{subfigure}[t]{.48\textwidth}
        \includesvg[width=\linewidth]{../results/clean_aov_height_residuals_cooks_dist}
        \caption{Cleansed data}
        \label{fig:clean_aov_height_residuals_cooks_dist}
    \end{subfigure}
    \caption{Cook's distance plot of the ANOVA residuals}
    \label{fig:aov_height_residuals_cooks_dist}
\end{figure}



Show the ANOVA table and an overview of the linkage and QTL plots you have generated in this course with the code you have used.

What are your major findings?

Are your traits in proximity to any of the following known genes and if so, can they explain your trait?

Locus Chromosome References:

VRS1 2H \url{https://doi.org/10.3390/ijms23042276}

ZEO1 2H \url{https://doi.org/10.1242/dev.170373}

Kap 4H \url{https://doi.org/10.1038/374727a0}

LKS2 7H \url{https://doi.org/10.1093/jxb/ers182}

Gpa1 2H \url{https://doi.org/10.1186/s12870-021-02915-9}

Nud 7H \url{https://doi.org/10.1073/pnas.0711034105}

Blp1 1H \url{https://doi.org/10.3389/fpls.2017.01414}
